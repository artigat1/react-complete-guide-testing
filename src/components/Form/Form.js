import React, { useEffect, useState } from 'react';

const TodoForm = props => {
    const [name, setName] = useState('');
    const [error, setError] = useState('');
    let textInput = null;
    let form = null;

    useEffect(() => textInput.focus());

    const handleSubmit = event => {
        event.preventDefault();

        if (validateForm()) {
            props.addItem({ newItemValue: name });
            form.reset();
            setName('');
            setError('');
        }
    };

    const handleChange = event => {
        event.preventDefault();
        validateForm();
    };

    const validateForm = () => {
        setError('');
        const value = textInput.value;
        setName(value);

        if (!value) {
            setError('You must enter a name');
            return false;
        }

        if (value.length <= 1) {
            setError('Name should be more than 1 characters');
            return false;
        }

        return true;
    };

    return (
        <form
            ref={f => {
                form = f;
            }}
            onSubmit={handleSubmit}
            noValidate
        >
            <input
                type="text"
                ref={input => {
                    textInput = input;
                }}
                placeholder="add a new todo..."
                onChange={handleChange}
                value={name}
            />
            <button type="submit" className="btn btn-default">
                Add
            </button>
            {error && <p className="error">{error}</p>}
        </form>
    );
};

export default TodoForm;
