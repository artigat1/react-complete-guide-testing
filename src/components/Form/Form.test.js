import React from 'react';
import renderer from 'react-test-renderer';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import Form from './Form';

configure({ adapter: new Adapter() });

describe('<Form />', () => {
    let wrapper;
    const fakeEvent = {
        preventDefault: () => jest.fn(),
        target: { value: 'testing' }
    };
    const mockAddItemCallback = jest.fn();

    beforeEach(() => {
        mockAddItemCallback.mockClear();

        wrapper = mount(<Form addItem={mockAddItemCallback} />);
    });

    it('should render correctly', () => {
        const component = renderer.create(<Form />).toJSON();
        expect(component).toMatchSnapshot();
    });

    it('should render submit button', () => {
        expect(wrapper.find('button[type="submit"]')).toHaveLength(1);
    });

    it('should have "Add" text on the submit button"', () => {
        const btnSubmit = wrapper.find('button[type="submit"]');

        expect(btnSubmit.text()).toEqual('Add');
    });

    it('should render input text box', () => {
        expect(wrapper.find('input[type="text"]')).toHaveLength(1);
    });

    it('should have a placeholder in the input box', () => {
        const inputTextBox = wrapper.find('input[type="text"]');

        expect(inputTextBox.prop('placeholder')).toEqual('add a new todo...');
    });

    it('should submit when valid text is entered in the name box', () => {
        const btnSubmit = wrapper.find('button[type="submit"]');
        const inputTextBox = wrapper.find('input[type="text"]');
        const expectedResult = { newItemValue: fakeEvent.target.value };

        inputTextBox.at(0).instance().value = fakeEvent.target.value;
        inputTextBox.simulate('change');
        wrapper.update();
        btnSubmit.simulate('submit', fakeEvent);
        wrapper.update();

        expect(mockAddItemCallback).toHaveBeenCalledTimes(1);
        expect(mockAddItemCallback).toHaveBeenCalledWith(expectedResult);
    });

    it('should show correct validation error when submitting no name', () => {
        const btnSubmit = wrapper.find('button[type="submit"]');
        const inputTextBox = wrapper.find('input[type="text"]');

        inputTextBox.at(0).instance().value = '';
        inputTextBox.simulate('change');
        wrapper.update();
        btnSubmit.simulate('submit', fakeEvent);
        wrapper.update();

        expect(wrapper.find('.error')).toHaveLength(1);
        expect(wrapper.find('.error').text()).toEqual('You must enter a name');
    });

    it('should show correct validation error when submitting short name', () => {
        const btnSubmit = wrapper.find('button[type="submit"]');
        const inputTextBox = wrapper.find('input[type="text"]');

        inputTextBox.at(0).instance().value = 'a';
        inputTextBox.simulate('change');
        wrapper.update();
        btnSubmit.simulate('submit', fakeEvent);
        wrapper.update();

        expect(wrapper.find('.error')).toHaveLength(1);
        expect(wrapper.find('.error').text()).toEqual(
            'Name should be more than 1 characters'
        );
    });
});
