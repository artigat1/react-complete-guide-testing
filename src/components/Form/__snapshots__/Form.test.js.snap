// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> should render correctly 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    placeholder="add a new todo..."
    type="text"
    value=""
  />
  <button
    className="btn btn-default"
    type="submit"
  >
    Add
  </button>
  
</form>
`;
